python/src/main/python/yaml-template/main.py

```diff
@@ -69,74 +60,12 @@ def _preparse_jinja_flags(argv):
   return pipeline_args
 
 
-def _configure_parser(argv):
-  parser = argparse.ArgumentParser()
-  parser.add_argument(
-    '--yaml_pipeline',
-    '--pipeline_spec',
-    help='A yaml description of the pipeline to run.')
-  parser.add_argument(
-    '--yaml_pipeline_file',
-    '--pipeline_spec_file',
-    help='A file containing a yaml description of the pipeline to run.')
-  parser.add_argument(
-    '--json_schema_validation',
-    default='generic',
-    help='none: do no pipeline validation against the schema; '
-         'generic: validate the pipeline shape, but not individual transforms; '
-         'per_transform: also validate the config of known transforms')
-  parser.add_argument(
-    '--jinja_variables',
-    default=None,
-    type=json.loads,
-    help='A json dict of variables used when invoking the jinja preprocessor '
-         'on the provided yaml pipeline.')
-  return parser.parse_known_args(argv)
-
-
-def _pipeline_spec_from_args(known_args):
-  if known_args.yaml_pipeline_file and known_args.yaml_pipeline:
-    raise ValueError(
-      "Exactly one of yaml_pipeline or yaml_pipeline_file must be set.")
-  elif known_args.yaml_pipeline_file:
-    with FileSystems.open(known_args.yaml_pipeline_file) as fin:
-      pipeline_yaml = fin.read().decode()
-  elif known_args.yaml_pipeline:
-    pipeline_yaml = known_args.yaml_pipeline
-  else:
-    raise ValueError(
-      "Exactly one of yaml_pipeline or yaml_pipeline_file must be set.")
-
-  return pipeline_yaml
-
-
-class _BeamFileIOLoader(jinja2.BaseLoader):
-  def get_source(self, environment, path):
-    source = FileSystems.open(path).read().decode()
-    return source, path, lambda: True
-
-
 def run(argv=None):
-  argv = _preparse_jinja_flags(argv)
-  known_args, pipeline_args = _configure_parser(argv)
-  pipeline_yaml = (  # keep formatting
-    jinja2.Environment(
-      undefined=jinja2.StrictUndefined, loader=_BeamFileIOLoader())
-    .from_string(_pipeline_spec_from_args(known_args))
-    .render(**known_args.jinja_variables or {}))
-  pipeline_spec = yaml.load(pipeline_yaml, Loader=yaml_transform.SafeLineLoader)
-
-  with beam.Pipeline(  # linebreak for better yapf formatting
-          options=beam.options.pipeline_options.PipelineOptions(
-            pipeline_args,
-            pickle_library='cloudpickle',
-            **yaml_transform.SafeLineLoader.strip_metadata(pipeline_spec.get(
-              'options', {}))),
-          display_data={'yaml': pipeline_yaml}) as p:
-    print("Building pipeline...")
-    yaml_transform.expand_pipeline(
-      p, pipeline_spec, validate_schema=known_args.json_schema_validation)
-    print("Running pipeline...")
+  parser = argparse.ArgumentParser()
+  _, pipeline_args = parser.parse_known_args(_preparse_jinja_flags(argv))
+  pipeline_args += ['--sdk_location=container']
+  cache_provider_artifacts.cache_provider_artifacts()
```

Comment:
This should be called at container build time, not at runtime. (In particular, it will attempt to cache all artifacts, whether or not they're used.

Reply:
How can it be called at runtime? The template is not expanded during container build time. Building the container just involves running the Dockerfile that creates the launcher image by taking the base image and copying in relevant files like the template file itself (main.py). 

This would get run during expansion in the template launcher which I thought then cached artifacts for the workers to use

Reply:
I was suggesting that artifact caching be done at container build time, via a RUN command in the docker file. 

`cache_provider_artifacts.cache_provider_artifacts` installs the artifacts into the current environment (which is the container the template launcher uses and will not impact or help the workers).

Reply:
Ah ok after looking at the source code, I was misunderstanding what that function was actually doing. I will add that logic to the Dockerfile in this PR

Reply:
This reduced startup by ~30 seconds, so thanks for the call out!